File: leetcode/0697-degree-of-an-array/findShortestSubArray.ts
function findShortestSubArray(nums: number[]): number {
  // [出现次数, 左值, 右值]
  const map = new Map<number, [number, number, number]>();

  for (const [index, num] of nums.entries()) {
    if (map.has(num)) {
      const arr = map.get(num)!;
      arr[0]++;
      arr[2] = index;
      continue;
    }

    map.set(num, [1, index, index]);
  }

  let minLength = 0;
  let maxCount = 0;

  for (const [count, left, right] of map.values()) {
    const length = right - left + 1;
    if (maxCount < count) {
      maxCount = count;
      minLength = length;
      continue;
    }

    // 判断
    if (count < maxCount) {
      continue;
    }
    if (length > minLength) {
      continue;
    }
    minLength = length;
  }

  return minLength;
}

export default findShortestSubArray;

File: leetcode/0697-degree-of-an-array/findShortestSubArray.go
package leetcode

func findShortestSubArray(nums []int) int {
	// [出现次数, 左值, 右值]
	numMap := map[int][]int{}

	for index, num := range nums {
		if arr, exists := numMap[num]; exists {
			arr[0]++
			arr[2] = index
			continue
		}

		numMap[num] = []int{1, index, index}
	}

	var minLength = 0
	var maxCount = 0

	for _, v := range numMap {
		var (
			count  = v[0]
			left   = v[1]
			right  = v[2]
			length = right - left + 1
		)

		if maxCount < count {
			maxCount = count
			minLength = length
			continue
		}
		// 判断
		if count < maxCount {
			continue
		}
		if length > minLength {
			continue
		}
		minLength = length
	}

	return minLength
}
